### ISSUE TITLE
CURVE Auth missing backwards compatibility ?

### ISSUE THREAD

**Issue by TMesot** [zeromq/libzmq#1551]
Hello, 
I've experienced some strange issues with libzmq 4.x and curve authentication.
After 127 working REQ-REP messages exchanges, the client and the server hang and nothing more appends.

After a few tests it seems there is no backwards compatibility between 4.0.X and 4.1.X versions. Is it normal and is it already documented ?

Here are the test case with the result, the test sources codes and the test VM configuration. If needed I also have the wireshark capture and VMs are still available for testing. 

**Result of tests**
``` 

From : 4.0.5 to 4.0.5     : OK (local)
From : 4.0.5 to 4.1.1	  : NOK Hang after 126 REQ/REP Sending 127 REQ, NO REP
From : 4.0.5 to 4.1.3	  : NOK Hang after 126 REQ/REP Sending 127 REQ, NO REP
From : 4.0.5 to last git  : NOK Hang after 126 REQ/REP Sending 127 REQ, NO REP

From : 4.1.3 to 4.0.5	  : NOK Hang after 126 REQ/REP Sending 127 REQ, NO REP
From : 4.1.3 to 4.1.1     : OK
From : 4.1.3 to 4.1.3     : OK (local)
From : 4.1.3 to last git  : OK
```
**VMs configurations**
```
Debian 8.0 
Upgrade 21.08.15
libsodium from repository (sudo apt-get install libsodium-dev) 1.0.3

1 x zmq from 4.0.5  debian jessie repo(apt-get install libzmq3-dev) 
1 x zmq from 4.1.3 current stable sources 
1 x zmq from 4.1.1 http://download.zeromq.org/
1 x zmq last git commit/b07b1e270ecd5ef5d5afee25176669c28e59346c
```
**client.c**
```c
#include <zmq.h>
#include <zmq_utils.h>

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <assert.h>

//Client key - Pub : x>o4?OW+v]/8ydXnw)Stpi+}<N)Niw>j:YHm]!p/ Priv : Y#wdz+hb^m7}FVBi5vrBt9ph})a<ido)[K2cKpIn
//Server key - Pub : 3GvMWwxH0hO@ChUYIKxLq{*I}MMz.+an89UY3D3f Priv : y@idvi86zzM$<rk89V-{^+ngjh>fRyi?RcGsGk.6


int main (void)
{
    
    int major, minor, patch;
    zmq_version (&major, &minor, &patch); printf ("Current ØMQ version is %d.%d.%d\n", major, minor, patch);

    //Key stuff
    char public_key_c [41] = "x>o4?OW+v]/8ydXnw)Stpi+}<N)Niw>j:YHm]!p/";
    char secret_key_c [41] = "Y#wdz+hb^m7}FVBi5vrBt9ph})a<ido)[K2cKpIn";
    char public_key_s [41] = "3GvMWwxH0hO@ChUYIKxLq{*I}MMz.+an89UY3D3f";
    char secret_key_s [41] = "y@idvi86zzM$<rk89V-{^+ngjh>fRyi?RcGsGk.6";
    
    printf("Client key - Pub : %s Priv : %s\n", public_key_c, secret_key_c); 
    

    printf ("[Client]Connecting to hello world server…\n");
    void *context_c = zmq_ctx_new ();
    void *requester = zmq_socket (context_c, ZMQ_REQ);
    //Client curve setup
    int rc = zmq_setsockopt (requester, ZMQ_CURVE_SERVERKEY, public_key_s, 40);
    assert (rc == 0);
	rc = zmq_setsockopt (requester, ZMQ_CURVE_PUBLICKEY, public_key_c, 40);
    assert (rc == 0);
	rc = zmq_setsockopt (requester, ZMQ_CURVE_SECRETKEY, secret_key_c, 40);
    assert (rc == 0);
    zmq_connect (requester, "tcp://remote_ip:5555");
    int request_nbr; 
    for (request_nbr = 0; request_nbr != 2000; request_nbr++) {
        char buffer_c [10];
        printf ("[Client]Sending Hello %d…\n", request_nbr);
        zmq_send (requester, "Hello", 5, 0);

        zmq_recv (requester, buffer_c, 10, 0);
        printf ("[Client]Received World %d\n", request_nbr);

    }

    zmq_close (requester);

    zmq_ctx_destroy (context_c);

    return 0;
}
```
**server.c**
```c
#include <zmq.h>
#include <zmq_utils.h>

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <assert.h>


int main (void)
{
    //Key stuff
    
    char public_key_c [41] = "x>o4?OW+v]/8ydXnw)Stpi+}<N)Niw>j:YHm]!p/";
    char secret_key_c [41] = "Y#wdz+hb^m7}FVBi5vrBt9ph})a<ido)[K2cKpIn";
    char public_key_s [41] = "3GvMWwxH0hO@ChUYIKxLq{*I}MMz.+an89UY3D3f";
    char secret_key_s [41] = "y@idvi86zzM$<rk89V-{^+ngjh>fRyi?RcGsGk.6";
    
    printf("Server key - Pub : %s Priv : %s\n", public_key_s, secret_key_s); 
    //  Socket to talk to clients
    void *context_s = zmq_ctx_new ();
    void *responder = zmq_socket (context_s, ZMQ_REP);
    
    //Server curve setup
    int enable = 1; 
    int rc = zmq_setsockopt (responder, ZMQ_CURVE_SERVER, &enable, sizeof(int));
    assert (rc == 0);
    rc = zmq_setsockopt (responder, ZMQ_CURVE_SECRETKEY, secret_key_s, 40);
    assert (rc == 0);
    rc = zmq_bind (responder, "tcp://*:5555");
    assert (rc == 0);

    while(1){
        char buffer_s [10];
        zmq_recv (responder, buffer_s, 10, 0);
        printf ("[Server]Received Hello\n");
        sleep (1);          //  Do some 'work'
        zmq_send (responder, "World", 5, 0);
    }
    zmq_close (responder);

    zmq_ctx_destroy (context_s);

    return 0;
}
```

**Comment by hintjens** [comment]
It's not normal and not documented. Yet there are vulnerabilities in 4.0's
curve code, and I'd upgrade to 4.1 in any case.

**Comment by xaqq** [comment]
It's possible that the CVE fixes (especially the patch that fixed replay
attack) broke compatibility since it is
now more strict regarding the nonce values.
If you want to test this, you can use a libzmq from git with a commit
before those fixes. Something like 2bf89bd0f866d6cf934b5bf47a776d67f37c359c
.

On Tue, Aug 25, 2015 at 5:40 PM, Pieter Hintjens <notifications@github.com>
wrote:

> It's not normal and not documented. Yet there are vulnerabilities in 4.0's
> curve code, and I'd upgrade to 4.1 in any case.
> On 25 Aug 2015 16:47, "TMesot" <notifications@github.com> wrote:
>
> > Hello,
> > I've experienced some strange issues with libzmq 4.x and curve
> > authentication.
> > After 127 working REQ-REP messages exchanges, the client and the server
> > hang and nothing more appends.
> >
> > After a few tests it seems there is no backwards compatibility between
> > 4.0.X and 4.1.X versions. Is it normal and is it already documented ?
> >
> > Here are the test case with the result, the test sources codes and the
> > test VM configuration. If needed I also have the wireshark capture and
> VMs
> > are still available for testing.
> >
> > *Result of tests*
> >
> >
> > From : 4.0.5 to 4.0.5 : OK (local)
> > From : 4.0.5 to 4.1.1 : NOK Hang after 126 REQ/REP Sending 127 REQ, NO
> REP
> > From : 4.0.5 to 4.1.3 : NOK Hang after 126 REQ/REP Sending 127 REQ, NO
> REP
> > From : 4.0.5 to last git : NOK Hang after 126 REQ/REP Sending 127 REQ,
> NO REP
> >
> > From : 4.1.3 to 4.0.5 : NOK Hang after 126 REQ/REP Sending 127 REQ, NO
> REP
> > From : 4.1.3 to 4.1.1 : OK
> > From : 4.1.3 to 4.1.3 : OK (local)
> > From : 4.1.3 to last git : OK
> >
> > *VMs configurations*
> >
> > Debian 8.0
> > Upgrade 21.08.15
> > libsodium from repository (sudo apt-get install libsodium-dev) 1.0.3
> >
> > 1 x zmq from 4.0.5 debian jessie repo(apt-get install libzmq3-dev)
> > 1 x zmq from 4.1.3 current stable sources
> > 1 x zmq from 4.1.1 http://download.zeromq.org/
> > 1 x zmq last git commit/b07b1e270ecd5ef5d5afee25176669c28e59346c
> >
> > *client.c*
>
> >
> > #include <zmq.h>
> > #include <zmq_utils.h>
> >
> > #include <stdio.h>
> > #include <unistd.h>
> > #include <string.h>
> > #include <assert.h>
> >
> > //Client key - Pub : x>o4?OW+v]/8ydXnw)Stpi+}<N)Niw>j:YHm]!p/ Priv :
> Y#wdz+hb^m7}FVBi5vrBt9ph})a<ido)[K2cKpIn
> > //Server key - Pub : 3GvMWwxH0hO@ChUYIKxLq{*I}MMz.+an89UY3D3f Priv :
> y@idvi86zzM$<rk89V-{^+ngjh>fRyi?RcGsGk.6
> >
> >
> > int main (void)
> > {
> >
> > int major, minor, patch;
> > zmq_version (&major, &minor, &patch); printf ("Current ØMQ version is
> %d.%d.%d\n", major, minor, patch);
> >
> > //Key stuff
> > char public_key_c [41] = "x>o4?OW+v]/8ydXnw)Stpi+}<N)Niw>j:YHm]!p/";
> > char secret_key_c [41] = "Y#wdz+hb^m7}FVBi5vrBt9ph})a<ido)[K2cKpIn";
> > char public_key_s [41] = "3GvMWwxH0hO@ChUYIKxLq{*I}MMz.+an89UY3D3f";
> > char secret_key_s [41] = "y@idvi86zzM$<rk89V-{^+ngjh>fRyi?RcGsGk.6";
> >
> > printf("Client key - Pub : %s Priv : %s\n", public_key_c, secret_key_c);
> >
> >
> > printf ("[Client]Connecting to hello world server…\n");
> > void *context_c = zmq_ctx_new ();
> > void *requester = zmq_socket (context_c, ZMQ_REQ);
> > //Client curve setup
> > int rc = zmq_setsockopt (requester, ZMQ_CURVE_SERVERKEY, public_key_s,
> 40);
> > assert (rc == 0);
> > rc = zmq_setsockopt (requester, ZMQ_CURVE_PUBLICKEY, public_key_c, 40);
> > assert (rc == 0);
> > rc = zmq_setsockopt (requester, ZMQ_CURVE_SECRETKEY, secret_key_c, 40);
> > assert (rc == 0);
> > zmq_connect (requester, "tcp://remote_ip:5555");
> > int request_nbr;
> > for (request_nbr = 0; request_nbr != 2000; request_nbr++) {
> > char buffer_c [10];
> > printf ("[Client]Sending Hello %d…\n", request_nbr);
> > zmq_send (requester, "Hello", 5, 0);
> >
> > zmq_recv (requester, buffer_c, 10, 0);
> > printf ("[Client]Received World %d\n", request_nbr);
> >
> > }
> >
> > zmq_close (requester);
> >
> > zmq_ctx_destroy (context_c);
> >
> > return 0;
> > }
> >
> > *server.c*
>
> >
> > #include <zmq.h>
> > #include <zmq_utils.h>
> >
> > #include <stdio.h>
> > #include <unistd.h>
> > #include <string.h>
> > #include <assert.h>
> >
> >
> > int main (void)
> > {
> > //Key stuff
> >
> > char public_key_c [41] = "x>o4?OW+v]/8ydXnw)Stpi+}<N)Niw>j:YHm]!p/";
> > char secret_key_c [41] = "Y#wdz+hb^m7}FVBi5vrBt9ph})a<ido)[K2cKpIn";
> > char public_key_s [41] = "3GvMWwxH0hO@ChUYIKxLq{*I}MMz.+an89UY3D3f";
> > char secret_key_s [41] = "y@idvi86zzM$<rk89V-{^+ngjh>fRyi?RcGsGk.6";
> >
> > printf("Server key - Pub : %s Priv : %s\n", public_key_s, secret_key_s);
> > // Socket to talk to clients
> > void *context_s = zmq_ctx_new ();
> > void *responder = zmq_socket (context_s, ZMQ_REP);
> >
> > //Server curve setup
> > int enable = 1;
> > int rc = zmq_setsockopt (responder, ZMQ_CURVE_SERVER, &enable,
> sizeof(int));
> > assert (rc == 0);
> > rc = zmq_setsockopt (responder, ZMQ_CURVE_SECRETKEY, secret_key_s, 40);
> > assert (rc == 0);
> > rc = zmq_bind (responder, "tcp://*:5555");
> > assert (rc == 0);
> >
> > while(1){
> > char buffer_s [10];
> > zmq_recv (responder, buffer_s, 10, 0);
> > printf ("[Server]Received Hello\n");
> > sleep (1); // Do some 'work'
> > zmq_send (responder, "World", 5, 0);
> > }
> > zmq_close (responder);
> >
> > zmq_ctx_destroy (context_s);
> >
> > return 0;
> > }
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/zeromq/libzmq/issues/1551>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zeromq/libzmq/issues/1551#issuecomment-134634056>.
>



-- 
Kapp Arnaud - Xaqq

**Comment by TMesot** [comment]
After some tests, its definitively the CVE fix regarding nonce values (0900a489213d74feb86fc0b343308fe7884a2a3c) that break compatibility. 
As the versions before this fix are vulnerable and the stable version is newer then this fix, the answer is to update libzmq in my project.
